contracts/taskModules/ProxyModule.sol

```diff
@@ -0,0 +1,84 @@
+// SPDX-License-Identifier: UNLICENSED
+
+pragma solidity ^0.8.12;
+
+import {GelatoBytes} from "../vendor/gelato/GelatoBytes.sol";
+import {TaskModuleBase} from "./TaskModuleBase.sol";
+import {LibDataTypes} from "../libraries/LibDataTypes.sol";
+import {IOpsUserProxy} from "../interfaces/IOpsUserProxy.sol";
+import {IOpsUserProxyFactory} from "../interfaces/IOpsUserProxyFactory.sol";
+
+contract ProxyModule is TaskModuleBase {
+    using GelatoBytes for bytes;
+
+    IOpsUserProxyFactory public immutable opsUserProxyFactory;
+
+    constructor(IOpsUserProxyFactory _opsUserProxyFactory) {
+        opsUserProxyFactory = _opsUserProxyFactory;
+    }
+
+    /**
+     * @inheritdoc TaskModuleBase
+     * @dev _taskCreator cannot create task to other user's proxy
+     */
+    function onCreateTask(
+        bytes32,
+        address _taskCreator,
+        address _execAddress,
+        bytes calldata,
+        bytes calldata
+    ) external override {
+        address proxy = _deployIfNoProxy(_taskCreator);
+
+        bool execToProxy = opsUserProxyFactory.isProxy(_execAddress);
+
+        if (execToProxy)
+            require(_execAddress == proxy, "ProxyModule: Only proxy owner");
+    }
+
+    /**
+     * @inheritdoc TaskModuleBase
+     * @dev _execData is encoded with proxy's `executeCall` function if
+     * _execAddress is not a proxy. If _execAddress is proxy, _execData
+     * should have function signatures of `executeCall` or `batchExecuteCall`
+     */
+    function preExecTask(
+        bytes32,
+        address _taskCreator,
+        address _execAddress,
+        bytes calldata _execData
+    ) external override returns (address, bytes memory execData) {
+        address proxy = _deployIfNoProxy(_taskCreator);
+
+        execData = _execAddress == proxy
```

Comment:
why do we need this part here? I think we can just assume that if users submit tasks specifying the proxy module, then they will simply always use the `execAddress` of the actual smart contract they want to execute. Furthermore, the data that will be generated off-chain at execution time by Gelato should simply be the payload for the actual `execAddress` and then we encode this data using the `_encodeWithOpsUserProxy` on-chain. 

Not sure why we have an `if` statement here. Can you explain @brandonchuah?

Reply:
it's a quirck to allow the user to encode batch execution through their proxy :
- normal user will pass raw exec data & the proxy module will encode them for the proxy,
- "advanced" user could directly encode the call using the proxy batchExec

Not sure about the real use case we have for this though

Reply:
Yes. What @goums said was the idea